### DISCUSSION TITLE
Cannot trigger geo-layers onHover event with cypress mouseover

### DISCUSSION THREAD

**Discussion by convers39** [visgl/deck.gl#7310] [Q&A]
I am using Cypress to test the mouseover behavior, I have a `TileLayer` rendered on the DOM, and when hovering my mouse on the map, there is a tooltip popup beside the cursor (as below)

<img width="873" alt="image" src="https://user-images.githubusercontent.com/42597395/193709421-d924c69c-a793-478b-9fdf-038dc4c2c3c9.png">

<img width="1186" alt="image" src="https://user-images.githubusercontent.com/42597395/193710834-b4795a51-4e6d-48ee-944e-85235bb5034d.png">

Now when I try to test this behavior with Cypress, the Cypress mouseover event seems not to work with the `onHover` function in `TileLayer`, not sure what is the reason here. I have tried both the official simulated mouseover with `cy.get('element').trigger('mouseover')` and the [real event](https://github.com/dmtrKovalenko/cypress-real-events#cyrealhover), none of them worked.

```ts
  it('should show coords when hovering over', () => {
        // move mouse to the map center
        cy.get('canvas[class="mapboxgl-canvas"]').realHover({
          position: 'center',
        })
        // cy.get('canvas[class="mapboxgl-canvas"]').realHover({position: 'top'})
        // cy.get('canvas[class="mapboxgl-canvas"]').realHover({
        //   position: 'bottom',
        // })
        // cy.get('canvas[class="mapboxgl-canvas"]').realHover({position: 'left'})
        // cy.get('canvas[class="mapboxgl-canvas"]').realHover({position: 'right'})
        // cy.get('div[id="view-default-view"]').realHover()

        // verify tooltips exist
        cy.findByTestId('layerTooltip').should('exist')
        cy.findByTestId('geoTooltip').should('exist')
      })
  })

```

A slight difference between using Cypress trigger and real event is that, the Cypress trigger can only work when select the `div[id="view-default-view"` element, if I try to select the canvas then it will fail as the canvas is covered by the div. With real event I don't have this issue. While both didn't trigger the `onHover` function.

```tsx
import { TileLayer } from "@deck.gl/geo-layers";
import { StaticMap } from "react-map-gl";

// ...
    const onHoverTileLayer = (info: DeckglTileLayerInformation) => {
        console.log("onHoverTileLayer triggered"); // did not print to console when testing with Cypress
        onHoverLayer(info);

    };
// ...
    const layer = new TileLayer({
        pickable: true,
        onHover: onHoverTileLayer,
     });
// ...
     return (
            <DeckGL
                onViewStateChange={onViewStateChange}
                viewState={viewState}
                layers={[...layers, layer]}
                controller={true}
            >
                <StaticMap
                    width={"100%"}
                    height={"100%"}
                    mapStyle={`mapbox://styles/mapbox/${mapStyle}`}
                    mapboxApiAccessToken={MAPBOX_ACCESS_TOKEN}
                />
            </DeckGL>
     )
```

A strange thing is that when I try to pin on a step, such as a step of getting the canvas element in Cypress, the map is not showing (as below).

![image](https://user-images.githubusercontent.com/42597395/193711513-801512f6-f416-4c1c-894e-d2b020ab1a1e.png)

If I remove the pin then it shows the map again, and the `onHover` function is triggered normally when I move the cursor in.
![image](https://user-images.githubusercontent.com/42597395/193711973-218432fb-defe-4309-9408-dc118b479885.png)

Does anyone have any clues on this issue?

**Comment by Pessimistress** [comment]
The event manager of deck.gl listens to `pointermove` for hover events.

**Reply by convers39** [reply]
Thank you for your reply.

I have tried `mouseover`, `mousedown`, `pointermove`, `mousemove` but none of them works.

It is strange that when I pin the step I can see the pointer in the middle but the map is missing.

<img width="1693" alt="Screen Shot 2022-10-06 at 9 17 05" src="https://user-images.githubusercontent.com/42597395/194186415-679f1550-0622-4b8e-9908-12d632ef43fb.png">

**Reply by Pessimistress** [reply]
You need to wait for 1) the WebGL context to be ready 2) the TileLayer to fully load before trying to trigger any events. To do this check `layer.isLoaded` in the `onAfterRender` callback.

**Reply by convers39** [reply]
Thank you. 

I think the map should be rendered before the mouse event (if the order logs properly), but not sure where to find the `layer.isLoaded` property.

<img width="1110" alt="image" src="https://user-images.githubusercontent.com/42597395/194208051-dc2692d2-2106-4e06-8118-a8a8b447db62.png">

<img width="1847" alt="image" src="https://user-images.githubusercontent.com/42597395/194209273-92d93648-0d6c-4b9e-ac6d-32594f022718.png">

```ts
        cy.get('canvas[id="deckgl-overlay"]').then($canvas => {
          const canvasHeight = $canvas.height()
          const canvasWidth = $canvas.width()
          const centerX = canvasWidth / 2
          const centerY = canvasHeight / 2

          console.log('mouse event start')
          cy.wrap($canvas)
            .scrollIntoView()
            .click(centerX, centerY, {force: true})
            .trigger('pointermove', {
              force: true,
              clientX: centerX,
              clientY: centerY,
            })
            .trigger('mousemove', {
              force: true,
              clientX: centerY,
              clientY: centerY,
            })
            .trigger('mousedown', {
              force: true,
              pageY: centerY + 50,
              pageX: centerX + 50,
            })
            .realMouseDown({position: {x: centerX, y: centerY}})
            .realMouseMove(centerX, centerY)
            .realHover({position: {x: centerX, y: centerY}})
            .trigger('pointerdown', {
              force: true,
              x: 450,
              y: 300,
              isPrimary: true,
            })
            .trigger('pointermove', {x: 380, y: 240, force: true})
            .trigger('pointerover', {x: 380, y: 240, force: true})

          // verify tooltips exist
          cy.findByTestId('layerTooltip', {timeout: 5000}).should('exist')
        })
      })

```

I found similar issues on the cypress side, but no related solution...

https://github.com/cypress-io/cypress/issues/20865
https://github.com/cypress-io/cypress/issues/6875

**Comment by convers39** [comment]
I think the event is triggered but somehow the `onHover` callback didnt run

```ts
        cy.get('canvas[id="deckgl-overlay"]').then(
          ($canvas: JQuery<HTMLCanvasElement>) => {
            const ctx = $canvas[0].getContext('webgl2')
            console.log('canvas context: ', ctx)
            $canvas.on('pointerover', el => {
              console.log('pointerover: ', el)
            })
            $canvas.on('mouseover', el => {
              console.log('mouseover: ', el)
            })
            $canvas.on('pointermove', el => {
              console.log('pointermove: ', el)
            })
            $canvas.on('mousemove', el => {
              console.log('mousemove: ', el)
            })
            const canvasHeight = $canvas.height()
            const canvasWidth = $canvas.width()
            const centerX = canvasWidth / 2
            const centerY = canvasHeight / 2

            console.log('mouse event start')
            cy.wrap($canvas)
              .scrollIntoView()
              .click(centerX, centerY, {force: true})
              .trigger('pointermove', {
                force: true,
                clientX: centerX,
                clientY: centerY,
              })
              .trigger('mousemove', {
                force: true,
                clientX: centerY,
                clientY: centerY,
              })
              .trigger('mousedown', {
                force: true,
                pageY: centerY + 50,
                pageX: centerX + 50,
              })
              .realMouseDown({position: {x: centerX, y: centerY}})
              .realMouseMove(centerX, centerY)
              .realHover({position: {x: centerX, y: centerY}})
              .trigger('pointerdown', {
                force: true,
                x: 450,
                y: 300,
                isPrimary: true,
              })
              .trigger('pointermove', {x: 380, y: 240, force: true})
              .trigger('pointerover', {x: 380, y: 240, force: true})

            // verify tooltips exist
            cy.findByTestId('layerTooltip', {timeout: 5000}).should('exist')
          },
        )
      })

```

<img width="1467" alt="image" src="https://user-images.githubusercontent.com/42597395/194232290-5186acb7-8f4f-4934-8f99-49320994862c.png">

```ts
import { TileLayer } from "@deck.gl/geo-layers";
import { StaticMap } from "react-map-gl";

// ...
    const onHoverTileLayer = (info: DeckglTileLayerInformation) => {
        console.log("onHoverTileLayer triggered"); // did not print to console when testing with Cypress
        onHoverLayer(info);

    };
// ...
    const layer = new TileLayer({
        pickable: true,
        onHover: onHoverTileLayer,
     });
// ...
     return (
            <DeckGL
                onViewStateChange={onViewStateChange}
                viewState={viewState}
                layers={[...layers, layer]}
                controller={true}
            >
                <StaticMap
                    width={"100%"}
                    height={"100%"}
                    mapStyle={`mapbox://styles/mapbox/${mapStyle}`}
                    mapboxApiAccessToken={MAPBOX_ACCESS_TOKEN}
                />
            </DeckGL>
     )
```

**Reply by convers39** [reply]
I couldn't find any events registered with `canvas#deckgl-overlay` element, which explains why the `onHover` didn't run. Does that mean the deckgl `onHover` event handler is not registered yet when I am running this test, or registered to a different element?

<img width="1514" alt="image" src="https://user-images.githubusercontent.com/42597395/194236262-e812d744-d735-405f-bf4e-47e456e1914f.png">

```ts
            // @ts-ignore
            const events = Cypress.$._data($canvas[0], 'events')
            console.log('events registered before: ', events)
            $canvas.on('pointerover', el => {
              console.log('pointerover: ', el)
            })
            $canvas.on('mouseover', el => {
              console.log('mouseover: ', el)
            })
            $canvas.on('pointermove', el => {
              console.log('pointermove: ', el)
            })
            $canvas.on('mousemove', el => {
              console.log('mousemove: ', el)
            })
            // @ts-ignore
            const events = Cypress.$._data($canvas[0], 'events')
            console.log('events registered after: ', events)

```

**Reply by Pessimistress** [reply]
The event listener is registered on the parent div of the canvas in the React case.

**Reply by convers39** [reply]
yeah, I was wondering that and also checked the `div[id=deckgl-wrapper]` but got the same undefined..

<img width="1419" alt="image" src="https://user-images.githubusercontent.com/42597395/194439946-65973843-376c-48ba-9a9e-1b38edacd1ed.png">

**Reply by convers39** [reply]
@Pessimistress Hi, I tried with the `realHover` api from [cypress-real-events plugin](https://github.com/dmtrKovalenko/cypress-real-events), and fire the hover event on random positions multiple times, it worked..

The cypress api `trigger('mouseover/mousedown/mousemove/pointerover/pointerdown/pointermove')` will not work.

Not sure why the registered events are empty tho.

Thank you for your advices and help, really appreciate it!!
